Use specialized intrinsics for `dot4{I, U}8Packed` on SPIR-V and HLSL

robamler opened the pull request:
**Connections**
- Following @teoxoy's comment in https://github.com/gfx-rs/wgpu/pull/7494#issuecomment-2794505724

**Description**

Replace the polyfills for `dot4I8Packed` and `dot4U8Packed` from #7494 with specialized instructions where they are available. More precisely:

- **For HLSL,** we check if Shader Model >= 6.4, in which case we use `dot4add_{i,u}8packed` (the [specification](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/hlsl-shader-model-6-4-features-for-direct3d-12) explicitly states: "no separate capability bit check is required, beyond assuring the use of Shader Model 6.4")
- **For SPIR-V,** we check if the capabilities `DotProduct` and `DotProductInput4x8BitPacked` are available, in which case we use the "Packed Vector Format" argument for [`Op{S,U}Dot`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSDot) (and we emit the corresponding `OpCapability` and `OpExtension` statements at the beginning of the output).

If either of these tests fail, or if we are on Metal or GLSL, we fall back to the existing polyfills.

**Testing**

I added two tests with `.toml` configuration files that explicitly enable/prevent these specializations.

**Squash or Rebase?**

I think each commit should pass all CI tests.

**Open Questions / Notes**

- [x] ~On SPIR-V, the capabilities required for this specialization are only defined if `lang_version >= 1.6`, but I didn't check for high enough `lang_version` because I already check the `capabilities_available` field. This follows precedent in other places in the code base, but I'm not sure if it's sufficient to only check `capabilities_available`. If `capabilities_available == None`, then comments in the code ([here](https://github.com/gfx-rs/wgpu/blob/5304c3ca4a43e78d6a8e3b45d8d3f76216f2c1a6/naga/src/back/spv/mod.rs#L728) and [here](https://github.com/gfx-rs/wgpu/blob/5304c3ca4a43e78d6a8e3b45d8d3f76216f2c1a6/naga/src/back/spv/mod.rs#L840)) indicate that this should be interpreted as "all capabilities are permitted", and so the specialized code will be generated. But I guess "all capabilities" is still restricted to "all capabilities that are defined for `lang_version`". Should I check the `lang_version` in this case somehow, or is `capabilities_available == None` only used for debugging anyway? Unfortunately, `lang_version` isn't exposed anymore at the stage where we do the check, so I'd have to add it as an extra field to the `Writer`.~
  - ~[Update: added a check in be9debd, but it's a bit hacky]~
  - [Update: I misunderstood this, see [this comment](https://github.com/gfx-rs/wgpu/pull/7574#issuecomment-2827381413); resolved in f44bd46d0.]
- [x] ~Unclear whether `OpSDot` with "Packed Vector Format" expects signed or unsigned arguments in spv (see [comment below](https://github.com/gfx-rs/wgpu/pull/7574#issuecomment-2816704471))~ [Update: resolved in [this comment](https://github.com/gfx-rs/wgpu/pull/7574#issuecomment-2816762335)]
- On Metal, we still fall back to the polyfill because, as far as I can tell, msl doesn't have builtin support for integer dot products (but it does support packed integer vectors, see `packed_char4` in Table 2.4 in the [specification](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)).

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

robamler committed:
Use intrinsics for `dot4{I, U}8Packed` on spv

robamler committed:
Use intrinsics for `dot4{I, U}8Packed` in HLSL

robamler committed:
Add intrinsics for `dot4{I,U}8Packed` to changelog

cwfitzgerald commented on naga/src/back/spv/block.rs:
```diff
@@ -1143,59 +1143,89 @@ impl BlockContext<'_> {
-                                    new_arg0_id,
-                                    arg0_id,
-                                ));
+                        if self.writer.lang_version() >= (1, 7)
```
Nit: 1.6 is the newest spirv spec, this should be (1, 6)

robamler replied:
Thanks!

robamler committed:
Check for spv language version

When checking for capabilities in SPIR-V,
`capabilities_available == None` indicates that all capabilities are
available. However, some capabilities are not even defined for all
language versions, so we still need to check if the requested
capabilities even exist in the language version we're using.

robamler commented:
> On SPIR-V, the capabilities required for this specialization are only defined if `lang_version >= 1.6`, but I didn't check for high enough `lang_version` because I already check the `capabilities_available` field. This follows precedent in other places in the code base, but I'm not sure if it's sufficient to only check capabilities_available. If `capabilities_available == None`, then comments in the code ([here](https://github.com/gfx-rs/wgpu/blob/5304c3ca4a43e78d6a8e3b45d8d3f76216f2c1a6/naga/src/back/spv/mod.rs#L728) and [here](https://github.com/gfx-rs/wgpu/blob/5304c3ca4a43e78d6a8e3b45d8d3f76216f2c1a6/naga/src/back/spv/mod.rs#L840)) indicate that this should be interpreted as "all capabilities are permitted", and so the specialized code will be generated. But I guess "all capabilities" is still restricted to "all capabilities that are defined for lang_version". Should I check the lang_version in this case somehow, or is `capabilities_available == None` only used for debugging anyway? Unfortunately, lang_version isn't exposed anymore at the stage where we do the check, so I'd have to add it as an extra field to the Writer.

In case the above text is confusing: I pushed be9debd as an example of how an additional check for `lang_version` could be implemented, but it's a bit hacky and I'm also not completely sure if this check is necessary (I don't know if `capabilities_available == None` is supposed to occur outside of tests).

robamler commented:
I realized that this implementation might not be correct: I'm unsure if [`OpSDot`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSDot) expects its arguments to be signed if "Packed Vector Format" is used. The current implementation always uses unsigned 32-bit integers for packed vectors, even if their 8-bit components are understood to be signed (that's [how WGSL does it](https://www.w3.org/TR/WGSL/#dot4I8Packed-builtin), and also how it feels natural to me). But the SPIR-V spec appears ambiguous to me in this regard, and SPIR-V sets a different precedent in [`OpBitFieldSExtract`](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpBitFieldSExtract), which expects its input to have same type as its result.

Does anyone with deeper understanding of SPIR-V know this? I asked a [question on StackOverflow](https://stackoverflow.com/questions/79582419/does-opsdot-in-spir-v-with-packed-vector-format-require-its-two-inputs-to-be) to solicit help.

robamler commented:
The sign issue has been resolved. The implementation should be correct as is. Turns out SPIR-V [doesn't care about signedness](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_unsigned_versus_signed_integers) in this case: "all signed and unsigned operations always work on unsigned types, and the semantics of operation come from the opcode" (credits to Nicol Bolas, see [this StackOverflow answer](https://stackoverflow.com/a/79582488/2422684)).

robamler commented:
I realized that, while this implementation works in `naga`, it can't actually be used yet by `wgpu` because it requires SPIR-V language version >= 1.6, but `wgpu-hal` [sets the language version to at most 1.3](https://github.com/gfx-rs/wgpu/blob/5573c573935b62b5c875e5512e79f23c2dd52b4a/wgpu-hal/src/vulkan/adapter.rs#L2015-L2021) when the user requests a device from an adapter. I'm expecting a similar issue for a DX12 backend.

This might become a common issue in the future whenever an operation can be optimized on a subset of backend versions/capabilities. I'm currently working on a solution by introducing a new feature to [`FeaturesWGPU`](https://docs.rs/wgpu/latest/wgpu/struct.FeaturesWGPU.html) for the optimization. Thus, if a user is interested in the optimization, they can check if the adapter supports it using `Adapter::features()`. If the adapter supports the optimization, the user can request it when calling `Adapter::request_device`. If the adapter does not support the optimization, then the user can still use their shader, just without requesting the optimization. Is this approach acceptable or am I overthinking this and there's an easier solution?

Current status: I'm trying to figure out how to query [`ash`](https://docs.rs/ash/latest/ash/) if the capability `DotProductInput4x8BitPacked` is supported. There seem to be a number of [related flags in `ash::vk`](https://docs.rs/ash/latest/ash/vk/struct.PhysicalDeviceShaderIntegerDotProductProperties.html#structfield.integer_dot_product8_bit_unsigned_accelerated), but I'm not sure how exactly they map to whether or not the capability `DotProductInput4x8BitPacked` is supported. The capability `DotProductInput4x8BitPacked` is defined in [`vk.xml`](https://raw.githubusercontent.com/KhronosGroup/Vulkan-Docs/refs/heads/main/xml/vk.xml), from which `ash:vk` is generated, but I'm still trying to figure out which identifiers in `ash:vk` are generated from this, if any.

robamler commented:
I separated out the [above issue](https://github.com/gfx-rs/wgpu/pull/7574#issuecomment-2819248454) of how these naga optimizations can be used in `wgpu` to #7595. I think the present PR can be reviewed independently from #7595.

robamler commented:
Sorry for the back-and-forth on this one. I realized that I think I misunderstood the interplay between SPIR-V versions, SPIR-V extensions, and vulkan device extensions. My current understanding is:
- The vulkan device extension [VK_KHR_shader_integer_dot_product](https://registry.khronos.org/vulkan/specs/latest/man/html/VK_KHR_shader_integer_dot_product.html) has a "dependency" on the SPIR-V extension [SPV_KHR_integer_dot_product](https://github.khronos.org/SPIRV-Registry/extensions/KHR/SPV_KHR_integer_dot_product.html).
- The latter _requires only SPIR-V 1.0_, and it defines ops and capabilities with the "KHR" suffix (e.g., `OpSDotKHR` and `DotProductInput4x8BitPackedKHR`).
- These ops and capabilities seem to be integrated into SPIR-V proper (i.e., without requiring an extension) as of SPIR-V version 1.6 (see, e.g., right column [here](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpSDot)); here, their "KHR" suffix may be dropped (but may also still be used for backward compatibility).

Thus, as far as I understand now, the specialized instructions required for this optimization to work on SPIR-V should be available if we're either
- on SPIR-V 1.6; or
- on an older version of SPIR-V, and we requested the SPIR-V extension SPV_KHR_integer_dot_product, which we can do on older versions of SPIR-V as long as the vulkan adapter has device extension VK_KHR_shader_integer_dot_product.

In either case, we still have to check if the capabilities `DotProductKHR` and `DotProductInput4x8BitPackedKHR` are available (and request them).

Is my understanding correct? I'd appreciate any corrections because I want to make sure I don't introduce any regressions on exotic platforms.

teoxoy commented:
> Is my understanding correct? I'd appreciate any corrections because I want to make sure I don't introduce any regressions on exotic platforms.

That is correct. A few pointers in this direction: `Writer::capabilities_available` contains the capabilities passed by wgpu-hal so, it's enough to consult if the `DotProduct` & `DotProductInput4x8BitPacked` capabilities are present in `capabilities_available`, then based on that emit the fast path. Additionally, if the capabilities are present but we are emitting a version of SPIR-V lower than 1.6, only then call `use_extension("SPV_KHR_integer_dot_product")`.

teoxoy commented:
> I don't know if `capabilities_available == None` is supposed to occur outside of tests

If `capabilities_available` is empty, that should really mean that no extra capabilities can be used.
Snapshot tests can also specify `capabilities` if needed (via the toml).

teoxoy commented:
> - On Metal, we still fall back to the polyfill because, as far as I can tell, msl doesn't have builtin support for integer dot products (but it does support packed integer vectors, see `packed_char4` in Table 2.4 in the [specification](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf)).

It could be that if we use packed vectors as in https://github.com/gpuweb/gpuweb/issues/2677#issuecomment-1713292226 (or `char`` directly), the Metal compiler will compile it down to more specialized instructions.

robamler committed:
Optimize `dot4{I, U}8Packed` for all spv versions

Emit optimized code for `dot4{I, U}8Packed` regardless of SPIR-V version
as long as the required capabilities are available. On SPIR-V < 1.6,
require the extension "SPV_KHR_integer_dot_product" for this. On
SPIR-V >= 1.6, don't require the extension because the corresponding
capabilities are part of SPIR-V >= 1.6 proper.

robamler commented:
Thank you so much for the detailed feedback!

> [...] it's enough to consult if the `DotProduct` & `DotProductInput4x8BitPacked` capabilities are present in `capabilities_available`, then based on that emit the fast path. Additionally, if the capabilities are present but we are emitting a version of SPIR-V lower than 1.6, only then call `use_extension("SPV_KHR_integer_dot_product")`.

Done (in f44bd46).

> It could be that if we use packed vectors as in https://github.com/gpuweb/gpuweb/issues/2677#issuecomment-1713292226 (or `char` directly), the Metal compiler will compile it down to more specialized instructions.

Thanks for the pointer! I implemented this in 848def1, but validating the generated msl code leads to error messages like "error: as_type cast from 'unsigned int' to 'packed_uchar4' (packed vector of 4 'unsigned char' values) is not allowed" (see [CI output](https://github.com/gfx-rs/wgpu/actions/runs/14695539412/job/41236634386?pr=7574#step:5:35)). I'm surprised by this since the [MSL specification](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf) states in Section 2.20:

> Metal adds an `as_type<type-id>` operator to allow any scalar or vector data type (that is not a pointer) to be reinterpreted as another scalar or vector data type of the same size.

And Table 2.4 in Section 2.2.3 states the size of `packed_char4` as 4 bytes (i.e., same as `uint`; it does have lower alignment requirements than `uint`, but that should only be an issue for the opposite conversion, i.e., from `packed_char4` to `uint`, which we don't want to do here). Maybe `packed_char4` doesn't count as a "scalar or vector data type"?

I've removed this commit from the PR for now (so Metal falls back again to the original polyfill).

teoxoy commented:
It seems to only work in Metal 2.1+ https://shader-playground.timjones.io/6f0764f9a4538417e52931dd96849619, but I can't find this being mentioned in their spec. We can land this in a separate PR though since I think this one looks good already.
